### ISSUE TITLE
Unable to debug or hit breakpoints in Startup.cs

### ISSUE THREAD

**Issue by TroyWitthoeft** [OmniSharp/omnisharp-vscode#3644]
## Issue Description ##

Unable to debug or hit breakpoints in a Startup.cs file used by an Azure Functions project in VS Code.  Breakpoints in other files are hit.  The same project when when debugged in Visual Studio is able to hit breakpoints in the Startup.cs file.     

I have created a repro repo here.   https://github.com/TroyWitthoeft/AzureFunctionsHostStartupBug

Please clone down the repo and confirm that breakpoints in Startup.cs are hit.  

## Steps to Reproduce ##

1. Clone repo https://github.com/TroyWitthoeft/AzureFunctionsHostStartupBug
2. Launch VS Code and open project.
3. Place a breakpoint in Startup.cs before line 14.
4. Start debugger.

## Expected Behavior ##

Breakpoint is hit.  Exception is not thrown. 

## Actual Behavior ##

Breakpoint is not hit.  Exception is thrown. 
### OmniSharp log ###

Starting OmniSharp server at 3/9/2020, 9:09:09 AM
    Target: c:\Projects\StartupBug

OmniSharp server started.
    Path: C:\Users\TWitthoeft\.vscode\extensions\ms-vscode.csharp-1.21.12\.omnisharp\1.34.13\OmniSharp.exe
    PID: 22704

[info]: OmniSharp.Stdio.Host
        Starting OmniSharp on Windows 6.2.9200.0 (x64)
[info]: OmniSharp.Services.DotNetCliService
        DotNetPath set to dotnet
[info]: OmniSharp.MSBuild.Discovery.MSBuildLocator
        Located 2 MSBuild instance(s)
            1: Visual Studio Community 2019 16.4.29728.190 - "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Current\Bin"
            2: StandAlone 16.4 - "C:\Users\TWitthoeft\.vscode\extensions\ms-vscode.csharp-1.21.12\.omnisharp\1.34.13\.msbuild\Current\Bin"
[info]: OmniSharp.MSBuild.Discovery.MSBuildLocator
        Registered MSBuild instance: Visual Studio Community 2019 16.4.29728.190 - "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Current\Bin"
[info]: OmniSharp.Cake.CakeProjectSystem
        Detecting Cake files in 'c:\Projects\StartupBug'.
[info]: OmniSharp.Cake.CakeProjectSystem
        Could not find any Cake files
[info]: OmniSharp.MSBuild.ProjectSystem
        No solution files found in 'c:\Projects\StartupBug'
[info]: OmniSharp.MSBuild.ProjectManager
        Queue project update for 'c:\Projects\StartupBug\StartupBug.csproj'
[info]: OmniSharp.Script.ScriptProjectSystem
        Detecting CSX files in 'c:\Projects\StartupBug'.
[info]: OmniSharp.Script.ScriptProjectSystem
        Could not find any CSX files
[info]: OmniSharp.WorkspaceInitializer
        Invoking Workspace Options Provider: OmniSharp.Roslyn.CSharp.Services.CSharpFormattingWorkspaceOptionsProvider, Order: 0
[info]: OmniSharp.MSBuild.ProjectManager
        Loading project: c:\Projects\StartupBug\StartupBug.csproj
[info]: OmniSharp.WorkspaceInitializer
        Invoking Workspace Options Provider: OmniSharp.Roslyn.CSharp.Services.RenameWorkspaceOptionsProvider, Order: 100
[info]: OmniSharp.WorkspaceInitializer
        Invoking Workspace Options Provider: OmniSharp.Roslyn.CSharp.Services.ImplementTypeWorkspaceOptionsProvider, Order: 110
[info]: OmniSharp.WorkspaceInitializer
        Invoking Workspace Options Provider: OmniSharp.Roslyn.CSharp.Services.EditorConfigWorkspaceOptionsProvider, Order: 200
[info]: OmniSharp.WorkspaceInitializer
        Configuration finished.
[info]: OmniSharp.Stdio.Host
        Omnisharp server running using Stdio at location 'c:\Projects\StartupBug' on host 24444.
[info]: OmniSharp.MSBuild.ProjectManager
        Successfully loaded project file 'c:\Projects\StartupBug\StartupBug.csproj'.
[info]: OmniSharp.MSBuild.ProjectManager
        Adding project 'c:\Projects\StartupBug\StartupBug.csproj'
[info]: OmniSharp.MSBuild.ProjectManager
        Update project: StartupBug

## Environment information ##

**VSCode version**: 1.42.1
**C# Extension**: 1.21.12

<details><summary>Dotnet Information</summary>
.NET Core SDK (reflecting any global.json):
 Version:   3.1.102
 Commit:    573d158fea

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.16299
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\3.1.102\

Host (useful for support):
  Version: 3.1.2
  Commit:  916b5cba26

.NET Core SDKs installed:
  3.1.101 [C:\Program Files\dotnet\sdk]
  3.1.102 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.15 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.15 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.15 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.2 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download
</details>
<details><summary>Visual Studio Code Extensions</summary>

|Extension|Author|Version|
|---|---|---|
|azure-account|ms-vscode|0.8.8|
|csharp|ms-vscode|1.21.12|
|vscode-azurefunctions|ms-azuretools|0.21.0|;

</details>

**Comment by WardenGnaw** [comment]
I have not used the Azure Functions extension much, but from the `.json` files, it looks like you are trying to debug a `Release` build of your project. 

The debugger has a difficult time guessing file line breakpoints with release symbols and its a random chance that it may hit a breakpoint.

Can you try building your project in `Debug` and if the breakpoint is still not being hit?

**Comment by cartermp** [comment]
Thanks for the report. I can't seem to reproduce the issue you're describing. After cloning your repo on my machine and installing the Azure Functions extension, the breakpoint is hit at line 13 (and also 14):

<img width="1093" alt="image" src="https://user-images.githubusercontent.com/6309070/76357687-63ca9b80-62d5-11ea-9667-99d6e8bb816d.png">

Can you confirm that this works after updating your C# and azure functions plugins?

**Comment by TroyWitthoeft** [comment]
@WardenGnaw 

On my machine the build process is only creating `bin/debug` and `obj/Debug` folders.  It's isn't creating any `Release` folders, so it certainly feels like we are creating files that should have all debug symbols.   

Here is my build command
`C:\Program Files\dotnet\dotnet.exe build /property:GenerateFullPaths=true /consoleloggerparameters:NoSummary`

Anything I can/should add in order to make sure the build process is creating debug symbols?   

@cartermp 

My `vscode-azurefunctions` plugin is current at v0.21.0.  And my `csharp` plugin is [intentionally](https://github.com/OmniSharp/omnisharp-vscode/issues/3623#issuecomment-595618372) one release behind at 1.21.12.   Trying to avoid conflating this with any of the other current issues that popped up in the 1.21.13 extension rename.   Thought that would be simpler.  However, I will upgrade and report back.  In the meantime, can you share your csharp and vscode-azurefunctions versions?  This very well may be an enviromental difference between our machines.  We'd need to get them close.

**Comment by cartermp** [comment]
@TroyWitthoeft Here's my info:

**C#:** 1.21.13
**vscode-azurefunctions:** 0.21.0

**tasks.json**

```json
{
    "version": "2.0.0",
    "tasks": [
        {
            "label": "clean",
            "command": "dotnet",
            "args": [
                "clean",
                "/property:GenerateFullPaths=true",
                "/consoleloggerparameters:NoSummary"
            ],
            "type": "process",
            "problemMatcher": "$msCompile"
        },
        {
            "label": "build",
            "command": "dotnet",
            "args": [
                "build",
                "/property:GenerateFullPaths=true",
                "/consoleloggerparameters:NoSummary"
            ],
            "type": "process",
            "dependsOn": "clean",
            "group": {
                "kind": "build",
                "isDefault": true
            },
            "problemMatcher": "$msCompile"
        },
        {
            "label": "clean release",
            "command": "dotnet",
            "args": [
                "clean",
                "--configuration",
                "Release",
                "/property:GenerateFullPaths=true",
                "/consoleloggerparameters:NoSummary"
            ],
            "type": "process",
            "problemMatcher": "$msCompile"
        },
        {
            "label": "publish",
            "command": "dotnet",
            "args": [
                "publish",
                "--configuration",
                "Release",
                "/property:GenerateFullPaths=true",
                "/consoleloggerparameters:NoSummary"
            ],
            "type": "process",
            "dependsOn": "clean release",
            "problemMatcher": "$msCompile"
        },
        {
            "type": "func",
            "dependsOn": "build",
            "options": {
                "cwd": "${workspaceFolder}/bin/Debug/netcoreapp3.1"
            },
            "command": "host start",
            "isBackground": true,
            "problemMatcher": "$func-watch"
        }
    ]
}
```

This is all just from cloning your repo.

**Comment by TroyWitthoeft** [comment]
I was able to reproduce this bug on another machine, but - fun times - it was transient! Sometimes I could hit the breakpoint. Othertimes, I could not.  What was the difference?  I was able to recognize a pattern.  On quick startups.  Startups where the `func.exe` quickly (1000-2000ms) loaded startup.cs file, the breakpoint was not hit.  If it took longer say 3000-6000ms then the startup.cs break point was hit! 

Have a look at the following gif.  
https://imgur.com/a/GE0Mn4A

I am running the same debug process.  On the first go round it takes four seconds to load startup.cs and the breakpoint is hit.  On the next run it takes two seconds and the breakpoint is NOT hit.  Is is possible this a timing issue?  How long does it take for the debugger to attach?  More than a second?  Is it possible that the debugger is unable to attach to func.exe before it loads the startup.cs file?  In your own sessions how many seconds elapse between  the first func.exe log message and the one where it indicates that startup.cs file has been loaded - `Loaded extension 'Startup' (1.0.0.0)`  ?   Are you able to get there in under a two seconds?

**Comment by TroyWitthoeft** [comment]
If I launch fast enough, it feels like I can get in and out of a func.exe session before the debugger can even identify the id of the running process.  I get this error.  

![Capture](https://user-images.githubusercontent.com/1102958/76420743-fb34fa80-6378-11ea-9a29-1b3a8abdf2d8.PNG)

**Comment by WardenGnaw** [comment]
Can you enable and share the logging for the debugger for the case where you do not hit the breakpoint? 

You can do this by adding the following to your `launch.json`:
```
"logging": {
   "engineLogging": true
}
```
Then we can figure out if the debugger is attaching too late if the breakpoint bound but never hit. 

If you are trying to hit a breakpoint that immediately happens during startup, it may be better to launch the process instead of attaching to it.

**Comment by cartermp** [comment]
Thanks for the detailed info @TroyWitthoeft - I removed the not-reproducible tag because it's clearly reproducing in your gif. At this point I'm inclined to degree there is a timing issue involved.

**Comment by TroyWitthoeft** [comment]
@WardenGnaw - It is enabled.  Line 10 of launch.json.  [Here](https://github.com/TroyWitthoeft/AzureFunctionsHostStartupBug/blob/c12c7dfe31fce7ca80fab7eb417f6a0333a60001/.vscode/launch.json#L10).  I have attached my debug console logs.
[DebugLogs.txt](https://github.com/OmniSharp/omnisharp-vscode/files/4320647/DebugLogs.txt).  

> if you are trying to hit a breakpoint that immediately happens during startup, it may be better to launch the process instead of attaching to it.

Okay, we can try that, but this is the default configuration.  Shouldn't it just work out of the box? i started down that path by reading the docs [here](https://code.visualstudio.com/docs/cpp/launch-json-reference) and [here](https://code.visualstudio.com/docs/editor/debugging#_launch-versus-attach-configurations) but I ran into this bug. 

I'm open to any suggestions on how to edit the `launch.json` to make that happen?   If we do make it work, is it possible that configuration be adopted as the standard? 

@cartermp -  Thanks.   Wanted to bring this to the attention of the team, because it makes for a not so great user experience.  I'm willing to dig into modifying the launch.json but I think it's fair to say that using the out-of-the-box configuration devs will expect all breakpoints to be hit.

**Comment by WardenGnaw** [comment]
@TroyWitthoeft Thanks for sharing your logs.

The debugger gets a pending breakpoint.
```
<- (R) {"seq":7,"type":"response","request_seq":4,"success":true,"command":"setBreakpoints","message":"","body":{"breakpoints":[{"id":2,"verified":false,"message":"The breakpoint is pending and will be resolved when debugging starts.","line":15},{"id":3,"verified":false,"message":"The breakpoint is pending and will be resolved when debugging starts.","line":16}]}}
```
But the breakpoint fails to bind since it can not find any symbols. 
```
<- (E) {"seq":15,"type":"event","event":"breakpoint","body":{"reason":"changed","breakpoint":{"id":2,"verified":false,"message":"No symbols have been loaded for this document.","line":15}}}
<- (E) {"seq":16,"type":"event","event":"breakpoint","body":{"reason":"changed","breakpoint":{"id":3,"verified":false,"message":"No symbols have been loaded for this document.","line":16}}}
```

The next step would be to see why the program's module is not being loaded. Can you also add `"moduleLoad": true` next to your engine logging?

e.g.
```
"logging": {
   "engineLogging": true,
   "moduleLoad": true
}
```

And share that output too?

---------

For setting up the `launch.json` for launch, You may want to contact the [Azure Functions extension](https://github.com/microsoft/vscode-azurefunctions/issues) to see if it is possible to use launch scenario for Azure Functions. 

They also own `${command:azureFunctions.pickProcess}`. So if this is a timing issue, they may be starting the program too fast before the debugger even has a chance to attach in time for that breakpoint. 

Another issue might be that it takes too long for them to get the pid before starting the debugger.

**Comment by TroyWitthoeft** [comment]
@WardenGnaw New Logs attached.   
[DebugLogs2.txt](https://github.com/OmniSharp/omnisharp-vscode/files/4325578/DebugLogs2.txt)

I've reached out to the Azure functions extension team with issue [1971](https://github.com/microsoft/vscode-azurefunctions/issues/1971)

**Comment by WardenGnaw** [comment]
Unfortunately, the provided logs (DebugLogs2.txt) did not successfully attach to the process so there was no chance for the debugger to load modules or debug breakpoints.

`<- (R) {"seq":10,"type":"response","request_seq":6,"success":false,"command":"configurationDone","message":"Failed to attach to process: Unknown Error: 0x8007012b"}`

From https://github.com/microsoft/vscode-azurefunctions/issues/1971, It looks like an issue on their end. 

I will be closing this as `Resolved-External`.

**Comment by TroyWitthoeft** [comment]
Thanks @WardenGnaw and @cartermp for helping diagnose.  Appreciate it.  Will follow up with the Azure functions extension team with issue [1971](https://github.com/microsoft/vscode-azurefunctions/issues/1971)
